src/prometheus/registry.zig

```diff
@@ -71,6 +72,75 @@ pub fn Registry(comptime options: RegistryOptions) type {
+            inline for (@typeInfo(Struct).Struct.fields) |field| {
+                try self.initMetric(Struct, &@field(metrics_struct, field.name), field.name);
+            }
+            return metrics_struct;
+        }
+
+        /// Initialize any fields within the struct that are supported metric types.
+        /// Leaves other fields untouched.
+        ///
+        /// Returns the number of fields that were *not* initialized.
+        pub fn initFields(
+            self: *Self,
+            /// Mutable pointer to a struct containing metrics.
+            metrics_struct: anytype,
+        ) GetMetricError!usize {
+            const Struct = @typeInfo(@TypeOf(metrics_struct)).Pointer.child;
+            const fields = @typeInfo(Struct).Struct.fields;
+            var num_fields_skipped: usize = fields.len;
+            inline for (@typeInfo(Struct).Struct.fields) |field| {
+                if (@typeInfo(field.type) == .Pointer) {
+                    const MetricType = @typeInfo(field.type).Pointer.child;
+                    if (@hasDecl(MetricType, "metric_type")) {
+                        try self.initMetric(Struct, &@field(metrics_struct, field.name), field.name);
+                        num_fields_skipped -= 1;
+                    }
+                }
+            }
+            return num_fields_skipped;
+        }
+
+        /// Assumes the metric type is **SomeMetric and initializes it.
+        ///
+        /// Uses the following declarations from Config:
+        /// - `prefix` if it exists
+        /// - `buckets` - required if metric is a *Histogram
+        ///
+        /// NOTE: does not support GaugeFn
+        ///
+        /// If expectations are violated, throws a compile error.
+        fn initMetric(
+            self: *Self,
+            Config: type,
+            /// Should be a mutable pointer to the location containing the
+            /// pointer to the metric, so `**SomeMetric` (ptr to a ptr)
+            metric: anytype,
+            comptime local_name: []const u8,
+        ) GetMetricError!void {
+            const MetricType = @typeInfo(@typeInfo(@TypeOf(metric)).Pointer.child).Pointer.child;
+            const prefix = if (@hasDecl(Config, "prefix")) Config.prefix ++ "_" else "";
+            const name = prefix ++ local_name;
+            metric.* = switch (MetricType.metric_type) {
+                .counter => try self.getOrCreateCounter(name),
+                .variant_counter => try self.getOrCreateVariantCounter(name, MetricType.Type),
+                .gauge => try self.getOrCreateGauge(name, MetricType.Data),
+                .gauge_fn => @compileError("GaugeFn does not support auto-init."),
+                .histogram => try self.getOrCreateHistogram(
+                    name,
+                    if (@typeInfo(@TypeOf(Config.buckets)) == .Fn)
+                        Config.buckets(name)
+                    else
+                        &Config.buckets,
```

Comment:
> maybe change this field name to default_buckets so its more clear that it will be used for all histograms

_First some context:_ Initially I had `Config.buckets(name)` as the only option, but I found that in many cases I was just implementing that function by ignoring the name and returning the same buckets no matter what. The function has to be written a bit strangely to ensure the comptime slice works correctly. I decided to allow it to be declared as a const for that common case, instead of the function, to reduce confusing boilerplate.

Normally I'd think of the word "default" to indicate that it's sort of like a fallback in the absence of explicit configuration. But in this case, you need to manually configure the buckets regardless. You need a `buckets` decl that can configure every single histogram in the struct. You just have the option to specify that configuration as being either dependent or independent of the field name. But you can't do both, this would make no sense. Since they are mutually exclusive, there's a benefit of using the same name for both, the compiler won't allow you to specify both by accident.

> is there a comptime way to ensure Config has either a buckets field or method 

Yes it currently checks this at comptime.

> and print a more helpful panic? (rn if its missing its not that great of an error message)

here's the compile error that we have now:
```zig
src/prometheus/registry.zig:136:49: error: struct 'ledger.writer.ScanAndFixRootsMetrics' has no member named 'buckets'
                    if (@typeInfo(@TypeOf(Config.buckets)) == .Fn)
                                          ~~~~~~^~~~~~~~
src/ledger/writer.zig:306:36: note: struct declared here
pub const ScanAndFixRootsMetrics = struct {
                                   ^~~~~~
referenced by:
    initStruct__anon_41729: src/prometheus/registry.zig:80:36
    test.setAndChainConnectedOnRootAndNextSlots: disconnected: src/ledger/writer.zig:475:62
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
```

Reply:
> -- rn the way we handle buckets is kind of hidden logic but not sure if theres a better way 🤔

I agree this could be better but I couldn't immediately think of a way to improve the situation. There is something nice about having explicit traits/interfaces built into the language.

Reply:
> here's the compile error that we have now:

yeah i more so meant can we have a better comptime error message -- one which describes how Metric structs which have histograms need to either provide a bucket field which will be used for all histograms or a bucket(field_name) method which returns the buckets for a given field name

usually i would be fine with keeping the compilers default compiler error but the bucket's usage is so hidden that a more descriptive comptime message would help devs debug a significant amount imo

Reply:
I like the idea of just having the simplest word "buckets" just be the canonical way to configure the histogram buckets

I agree a more descriptive error message is nice to have. I was just trying to keep the implementation simple, and I figured it was enough information to figure it out. It's significantly more complex to have custom error messages.

Also regarding the naming, I like something about just having the simplest and most obvious word "buckets" just be the canonical way to configure the histogram buckets, regardless of whether it's function or an array. I think "buckets" the easiest name to remember when writing the struct. If I kinda forgot exactly what name to use, "buckets" would probably be my first guess. You're less likely to see any comptime errors at all, compared to seeing an error and needing to have a nice error message, because your first guess would be correct. Whereas having different names with multiple words, it's a little harder to get it correct unless you see it written in front of you already. Also I felt that "buckets" is descriptive enough of a name to be pretty clear what it means. But I do also see the advantage of an explicit function name helping clarify the purpose of the string parameter.